### PR TITLE
Improve trivia handling in 'use auto property'.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/dotnet/roslyn/issues/26858

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
tagging @dotnet/roslyn-ide @sharwell @heejaechang @mavasani @jcouv 

Very simple tweak to improve some of the trivia handling in 'use auto prop'.  Thanks!

**Comment by Neme12** [comment]
Could we just move all trivia from the field to the property instead?

**Comment by CyrusNajmabadi** [comment]
> Could we just move all trivia from the field to the property instead?

The feeling, when this feature was created, was that would *not* be what people would want most of the time.  It would double-up trivia unnecessarily, and certainly would not be desirable for things like newlines.  i.e. if you had a newline on the field and a newline on hte property, you would not want to end up with two newlines on the final property.  Similarly, the comments for a field feel like they belong to that field.  When removing the field, more often than not the comments won't be appropriate to keep around (at best they will just be redundant).  So defaulting to just removing them produces the code the user intends the majority of the time.

In the cases it doesn't a manual fixup can be done instead.  However, it would be better to do the manual fixup more rarely than having it be something you often need to do.

File: src/EditorFeatures/VisualBasic/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb [hunk 1]
```diff
@@ -12,12 +11,26 @@ Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
...
     <ExportCodeFixProvider(LanguageNames.VisualBasic, Name:=NameOf(VisualBasicUseAutoPropertyCodeFixProvider)), [Shared]>
     Friend Class VisualBasicUseAutoPropertyCodeFixProvider
-        Inherits AbstractUseAutoPropertyCodeFixProvider(Of PropertyBlockSyntax, FieldDeclarationSyntax, ModifiedIdentifierSyntax, ConstructorBlockSyntax, ExpressionSyntax)
+        Inherits AbstractUseAutoPropertyCodeFixProvider(Of PropertyBlockSyntax, ModifiedIdentifierSyntax, ConstructorBlockSyntax, ExpressionSyntax)
 
         Protected Overrides Function GetNodeToRemove(identifier As ModifiedIdentifierSyntax) As SyntaxNode
             Return Utilities.GetNodeToRemove(identifier)
         End Function
 
+        Protected Overrides Function WillRemoveFirstFieldInTypeDirectlyAboveProperty(
```
[the remark this answers is not in the document]
Comment (Neme12) [reply src/EditorFeatures/VisualBasic/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb]: with a new type parameter for TypeBlockSyntax... or isn't there a helper for this in syntax facts?

[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasic/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb]: I'll take a look to see if possible.

File: src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs [hunk 2]
```diff
@@ -83,14 +86,33 @@ public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
...
                 isWrittenToOutsideOfConstructor, cancellationToken).ConfigureAwait(false);
 
+            // Ensure the new and old property share the same trailing trivia.
+            updatedProperty = updatedProperty.WithTrailingTrivia(property.GetTrailingTrivia());
+
+            // However, if we have a situation where the property is the second member in a type,
+            // and it would become the first, then remove any leading blank lines from it so we
+            // don't have random blanks above it that used to space it from the field that was
+            // there.
+            var nodeToRemove = GetNodeToRemove(declarator);
```
Comment (Neme12) [inline src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs]: why is GetNodeToRemove called twice? could you put this logic after the other call to GetNodeToRemove below?

Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/UseAutoProperty/AbstractUseAutoPropertyCodeFixProvider.cs]: Done.

File: src/EditorFeatures/VisualBasic/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb (hunk 1 above)
[the remark this answers is not in the document]
Comment (CyrusNajmabadi) [reply src/EditorFeatures/VisualBasic/UseAutoProperty/VisualBasicUseAutoPropertyCodeFixProvider.vb]: Done.

**Comment by CyrusNajmabadi** [comment]
@jcouv @jmarolf can i get some eyes?  Thanks!

**Comment by jcouv** [comment]
I'll look after lunch.

**Comment by CyrusNajmabadi** [comment]
> I'll look after lunch.

@jcouv  Thanks!

File: src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs [hunk 3]
```diff
@@ -1403,8 +1383,89 @@ protected set
...
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseAutoProperty)]
+        [WorkItem(26858, "https://github.com/dotnet/roslyn/issues/26858")]
+        public async Task TestPreserveTrailingTrivia1()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Goo
+{
+    private readonly object [|bar|] = new object();
+
+    public object Bar => bar;
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs]: >Bar [](start = 18, length = 3)

Not a new problem with this PR, but I don't see test scenarios where the property was defined above the field. Is that intentionally out-of-scope for the fixer?

File: src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs [hunk 4]
```diff
@@ -1403,8 +1383,89 @@ protected set
...
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseAutoProperty)]
+        [WorkItem(26858, "https://github.com/dotnet/roslyn/issues/26858")]
+        public async Task TestKeepLeadingBlank()
+        {
+            await TestInRegularAndScriptAsync(
+@"class Goo
+{
+
+    private readonly object [|bar|] = new object();
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs]: >bar [](start = 30, length = 3)

Would it make sense to test with some other fields or members above field `bar`?

**Comment by CyrusNajmabadi** [comment]
@jcouv  i will beef up the tests a bit!

**Comment by CyrusNajmabadi** [comment]
> Also, I'm unclear why "first member in type" is a special case (we should test the behavior with another member above the field+property being fixed).

Good question, and clearly i should comment things better.  The reason it is special is that the first member of a type tends to be special.  i.e. users do not normally put blank lines before the first member.  And so, when a type now *becomes* the first member, we want to follow the user's common pattern here.

In all other code cases, i.e. when tehre are multiple fields above, or the field is below the property, then then property isn't now becoming "the first member", and as such, it doesn't want this special behavior about it's leading blank lines.  i.e. if i have:

```c#
class C
{
    int i;
    int j;

    int Prop => j;
}
```

Then if i remove 'j' (or even 'i'), then 'Prop' would stay the non-first member, and would definitely *want* to keep that blank line above it.

In essence, the blank line above the property exists for separation from what's above it.  As long as something is above it, we keep the separation.  However, if the property becomes the *first* member in the type, the separation is now inappropriate because there's nothing ot actually separate it from.

Does that help clarify things?

**Comment by jcouv** [comment]
Makes sense. Thanks for the clarification.

File: src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs (hunk 3 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs]: Have added tests.

File: src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs]: Have added tests.

**Comment by mavasani** [comment]
@CyrusNajmabadi - The explanation was very helpful. However, I feel this issue is not specific to this particular code fixer, but applies to any fix/refactoring which attempts to add/remove/shuffle members. I wonder if we can move your fix to some common layer, such as Formatter, SyntaxEditor, SyntaxGenerator, or PostProcess on CodeActions or some such common infrastructure, so all code transformations get this by default. I don't suggest blocking this PR on such alternate approach, but something for us to consider.

Tagging @heejaechang @jinujoseph @JieCarolHu, who have been working on code cleanup feature design and thinking on approaches to ensure that invoking a code fix/refactoring performs an auto-cleanup on the changed code at the end.

**Comment by CyrusNajmabadi** [comment]
>  I wonder if we can move your fix to some common layer, such as Formatter, SyntaxEditor, SyntaxGenerator, or PostProcess on CodeActions or some such common infrastructure, so all code transformations get this by default

This is def an interesting idea.  I think my preference would be for this to be part of the RemoveNode API.  Either as part of the default behavior, or as an option that can go into RemoveOptions.  Effectively, we would say that if you do remove the first member in typedecl, and it had no leading blanks, then we would strip leading blanks on the member that now became the first.

File: src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs [hunk 5]
```diff
@@ -1403,8 +1383,189 @@ protected set
...
+@"class Class
+{
+    int j;
+
+    int P { get; }
+}");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsUseAutoProperty)]
+        public async Task TestMultipleFieldsAbove2()
```
Comment (jcouv) [inline src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs]: >TestMultipleFieldsAbove2 [](start = 26, length = 24)

Consider another variant of this test, where there is an empty line between `int j` and `int i`.
The question is what happens to the empty line above `int i` and the empty line between `int i` and the property for `i`. 

I'd expect we end up with a single empty line between `int j` (first field) and the property. Is that correct?

**Comment by heejaechang** [comment]
we already have something like this for members that got auto generated. 
http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Formatting/Rules/ElasticTriviaFormattingRule.cs,158

for code that doesn't have elastic trivia between 2 members, we have 2 choice.

1. add elastic trivia between 2 members that we don't want blank lines between, so that ElasticTriviaFormattingRules got called. if it is there, but not hitting the rule above, we should add new check in that rule above

at the end, what we need is returning between 2 members
FormattingOperations.CreateAdjustNewLinesOperation(1 /* 1 line difference between member */, AdjustNewLinesOption.ForceLines)

2. add custom rule and call format explicitly in the fixer.
this situation can happen if 2 fixer want to have 2 different behaviors. like 1 fixer (say, generate property) wants blank lines between property it generated and existing properties, 1 fixer (say, convert to expression body) wants no blank lines for property they converted. then there is no way to have 1 central rule since 2 fixers wants different behavior for same thing. in such situation, one should explicitly call formatter with custom rule.

...

that being said, for this particular case, I think all we need is make sure there is elastic trivia between member, and add new condition in 
http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Workspaces/Formatting/Rules/ElasticTriviaFormattingRule.cs,158

saying if 2 properties both have shape of expression body, just return 1 as lines between 2 members. and force it (AdjustNewLinesOption.ForceLines), so it always become 1 line.

**Comment by CyrusNajmabadi** [comment]
> that being said, for this particular case, I think all we need is make sure there is elastic trivia between member, and add new condition in

Note: in general we've avoided adding elastic trivia when unnecessary because it often has unintended consequences.  For example, we've seen at times, that it makes the entire member eligible for more aggressive formatting.  This is not what we want in certain cases because we're trying to preserve the user's code as closely as possible, and we only want to make very specific edits to it without affecting any other part of the tree.

File: src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs]: Will do!

**Comment by heejaechang** [comment]
if we are fine not putting this behavior in central way since it could be intrusive, you can do 
2. add custom rule and call format explicitly in the fixer.

create custom rule that return 
FormattingOperations.CreateAdjustNewLinesOperation(1, AdjustNewLinesOption.ForceLines) between 2 members you care about. 

and call Formatter.FormatAsync(... customeRule.Concat(Foramtter.GetDefaultRule()) ..)

or we can add extension method that has signature of
Document = FormateWithCustomRule(Document, ....., IFormattingRule) which does the above for you.

File: src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs (hunk 5 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/UseAutoProperty/UseAutoPropertyTests.cs]: @jcouv Let me know if the tests are sufficient :)

**Comment by CyrusNajmabadi** [comment]
> add custom rule and call format explicitly in the fixer.

Sure, i'm def not opposed to explicit formatting rules.  It's def something i've done in other cases.  :)

IN general, my litmus is simply this:  Can the feature get the formatting right triviall on it's own just by manipulating a small amount of trivia?  If so, great, just do it.  However, if it's going to be really difficult to manually work with trivia (for example, because we're generating entirely fresh code taht might go in any random indentation in the tree), then def just go with a formatting rule so we can benefit from having useful formatting functionality (especially like suppressions and anchoring rules).  Effectively, if you formatting tweaks are very dependent on the surrounding code, use an actual formatting rule.  If your formatting is effectively "include/exclude trivia" or do something incredibly basic, just leave it in the feature itself.

**Comment by heejaechang** [comment]
I am fine feature explicitly manipulating trivia to the shape they want. i am even fine one using ParseText to get that trivia shape rather than building them using factory.

I was just suggesting alternative in case you want to let formatter to do it.

**Comment by CyrusNajmabadi** [comment]
> I was just suggesting alternative in case you want to let formatter to do it.

Understood and appreciated.  Thanks!

**Comment by jcouv** [comment]
@jinujoseph for ask-mode approval for 15.8. Thanks

**Comment by jinujoseph** [comment]
approved for 15.8.preview3

**Comment by jinujoseph** [comment]
test windows_debug_vs-integration_prtest

**Pull request merged**